src/sinks/aws_cloudwatch_metrics.rs

```diff
@@ -0,0 +1,381 @@
+use crate::{
+    buffers::Acker,
+    event::{Event, Metric},
+    region::RegionOrEndpoint,
+    sinks::util::{
+        retries::{FixedRetryPolicy, RetryLogic},
+        BatchServiceSink, Partition, SinkExt,
+    },
+    topology::config::{DataType, SinkConfig},
+};
+use futures::{Future, Poll};
+use rusoto_cloudwatch::{
+    CloudWatch, CloudWatchClient, MetricDatum, PutMetricDataError, PutMetricDataInput,
+};
+use rusoto_core::{Region, RusotoFuture};
+use serde::{Deserialize, Serialize};
+use std::{convert::TryInto, time::Duration};
+use tower::{Service, ServiceBuilder};
+use tracing::field;
+use tracing_futures::{Instrument, Instrumented};
+
+#[derive(Clone)]
+pub struct CloudWatchMetricsService {
+    client: CloudWatchClient,
+    config: CloudWatchMetricsSinkConfig,
+}
+
+#[derive(Deserialize, Serialize, Debug, Clone, Default)]
+#[serde(deny_unknown_fields)]
+pub struct CloudWatchMetricsSinkConfig {
+    pub namespace: String,
+    #[serde(flatten)]
+    pub region: RegionOrEndpoint,
+    pub batch_size: Option<usize>,
+    pub batch_timeout: Option<u64>,
+
+    // Tower Request based configuration
+    pub request_in_flight_limit: Option<usize>,
+    pub request_timeout_secs: Option<u64>,
+    pub request_rate_limit_duration_secs: Option<u64>,
+    pub request_rate_limit_num: Option<u64>,
+    pub request_retry_attempts: Option<usize>,
+    pub request_retry_backoff_secs: Option<u64>,
+}
+
+#[typetag::serde(name = "aws_cloudwatch_metrics")]
+impl SinkConfig for CloudWatchMetricsSinkConfig {
+    fn build(&self, acker: Acker) -> Result<(super::RouterSink, super::Healthcheck), String> {
+        let sink = CloudWatchMetricsService::new(self.clone(), acker)?;
+        let healthcheck = CloudWatchMetricsService::healthcheck(self)?;
+        Ok((sink, healthcheck))
+    }
+
+    fn input_type(&self) -> DataType {
+        DataType::Metric
+    }
+}
+
+impl CloudWatchMetricsService {
+    pub fn new(
+        config: CloudWatchMetricsSinkConfig,
+        acker: Acker,
+    ) -> Result<super::RouterSink, String> {
+        let client = Self::create_client(config.region.clone().try_into()?)?;
+
+        let batch_size = config.batch_size.unwrap_or(5);
+        let batch_timeout = config.batch_timeout.unwrap_or(1);
+
+        let timeout = config.request_timeout_secs.unwrap_or(30);
+        let in_flight_limit = config.request_in_flight_limit.unwrap_or(5);
+        let rate_limit_duration = config.request_rate_limit_duration_secs.unwrap_or(1);
+        let rate_limit_num = config.request_rate_limit_num.unwrap_or(5);
```

Comment:
I'm thinking about bumping it to 25, what do you think? I'm a bit afraid of making Vector a 'busy neighbour' for possible other applications publishing metrics into CW, assuming this limit is applied on account level.

Reply:
Interesting, point, I do think though that this is configurable and we should probably default to the most that cloudwatch metrics supports. Then document that if you are using other applications and putting then you should lower it? What do you think?

Reply:
I agree with that. By the way, what should we do in case user is trying to provide configuration values beyond limits (especially those that cannot be changed)? Or we should leave it between user and AWS.

Reply:
I think in that case, we let them do it. They will get errors about being throttled so they should then be able to adjust.